### ISSUE TITLE
LAB-03: Clap Scores

### ISSUE THREAD

**Issue by Mazhar3204** [Watts-College/paf-514-template#38]
Hi Professor,

For the first question, I am not overthinking it. Do we look at the data set and view the claps scoring average it has, and then that would reveal which style title performs the best? Or are we supposed to create an expression to get the average? 

Q1-A: What is each type of title's average performance (clap score)?

Q1-B: Which style of title performs the best overall?

@lecy

**Comment by lecy** [comment]
You will first define the different types of titles as groups using grepl(), then you can calculate the average class score BY GROUP. Recall that the goal is to figure out the best style to maximize readership for your blog. 

Your solutions won't be identical because there are lots of decisions each person will make about operationalizing the group, but in the end they should look something like this: 

![image](https://github.com/Watts-College/paf-514-template/assets/1209099/f83819ad-ffd8-4753-ad26-2cd00e464792)

**Comment by jasminacosta** [comment]
> You will first define the different types of titles as groups using grepl(), then you can calculate the average class score BY GROUP. Recall that the goal is to figure out the best style to maximize readership for your blog.
> 
> Your solutions won't be identical because there are lots of decisions each person will make about operationalizing the group, but in the end they should look something like this:
> 
> ![image](https://private-user-images.githubusercontent.com/1209099/300060048-f83819ad-ffd8-4753-ad26-2cd00e464792.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDYyOTEwMTUsIm5iZiI6MTcwNjI5MDcxNSwicGF0aCI6Ii8xMjA5MDk5LzMwMDA2MDA0OC1mODM4MTlhZC1mZmQ4LTQ3NTMtYWQyNi0yY2QwMGU0NjQ3OTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjZUMTczODM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OGU1MDE4N2NhMzAwZWU2ZWU3MzdkZDkzODgxNjAzOTVlYzFkMWQ2MTNmOTY0MWFiNGQ4OTgyZDM3MmIyNmFjNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.d7SpuvSgHxdzPEyJWbiGgp1Zl4nZlYTDFAuH4jGGFSI)

@lecy I am not sure if I am understanding correctly. Am I going in the right direction? 

```{r}

power.lists <- grep("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
                    ^eight|^nine|^ten", d$title, value=TRUE, ignore.case=TRUE)
power.lists

power.lists.T.F <- grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
                    ^eight|^nine|^ten", d$title, ignore.case=TRUE)
power.lists.T.F 

power.list.data <- d[d$title %in% power.lists, c("title", "clap.score")]
mean(power.list.data$clap.score)


# Data Extraction 
power.list.score <- d %>%
  select(title, claps, clap.score) %>%
  filter(grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
                    ^eight|^nine|^ten", title, ignore.case=TRUE))

mean(power.list.score$claps)
mean(power.list.score$clap.score)

```

**Comment by lecy** [comment]
It looks correct. 

You can simplify the steps some: 

```r
clap.score <- log( d$clap + 1 )

power.list.T.F <- 
  grepl( "^[[:digit:]]|^one|^two|^three|^four|
          ^five|^six|^seven|^eight|^nine|^ten", 
          d$title, ignore.case=TRUE )

clap.score[ power.list.T.F ] %>% mean()
```

**Comment by jasminacosta** [comment]
@lecy 

That makes things a lot more easy.

Is it okay that I have been using my data from this part of my code? Because I have been running my functions using this, but it is the same outputs of data I was just trying to solve it in different ways. 

That's why I have power.list.T.F, power.list.score, and power.lists

```{r}
# Data Extraction 
power.list.score <- d %>%
  select(title, claps, clap.score) %>%
  filter(grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
                    ^eight|^nine|^ten", title, ignore.case=TRUE))

mean(power.list.score$claps)
mean(power.list.score$clap.score)
```

**Comment by lecy** [comment]
Absolutely. There are always lots of ways to arrive at the same answer in R. 

My code is always a mess (too long, very inefficient) when I'm working through a new problem. 

Just trying to show that you can often clean it up by eliminating the steps that were not needed. Practice writing readable code that is easy to maintain. 

For example, this is pretty dense and hard to follow: 

```r
power.list.score <- d %>%
  select(title, claps, clap.score) %>%
  filter(grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
                    ^eight|^nine|^ten", title, ignore.case=TRUE))
```

Simplify by separating the regular expression from the data wrangling step: 

```r
power.list <- grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
                    ^eight|^nine|^ten", title, ignore.case=TRUE)
d.power <- 
  d %>%
  select( title, claps, clap.score ) %>%
  filter( power.list )
```
You could take it a step further by isolating the messy regular expression: 

```r
regex.power <- 
  "^[[:digit:]]|^one|^two|^three|^four|
   ^five|^six|^seven|^eight|^nine|^ten"

power.list <- 
  grepl( regex.power, title, ignore.case=TRUE )

d.power <- 
  d %>%
  select( title, claps, clap.score ) %>%
  filter( power.list )
```

It makes each step easier to see, makes errors a lot easier to identify.

**Comment by jasminacosta** [comment]
> Absolutely. There are always lots of ways to arrive at the same answer in R.
> 
> My code is always a mess (too long, very inefficient) when I'm working through a new problem.
> 
> Just trying to show that you can often clean it up by eliminating the steps that were not needed. Practice writing readable code that is easy to maintain.
> 
> For example, this is pretty dense and hard to follow:
> 
> ```r
> power.list.score <- d %>%
>   select(title, claps, clap.score) %>%
>   filter(grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
>                     ^eight|^nine|^ten", title, ignore.case=TRUE))
> ```
> 
> Simplify by separating the regular expression from the data wrangling step:
> 
> ```r
> power.list <- grepl("^[[:digit:]]|^one|^two|^three|^four|^five|^six|^seven|
>                     ^eight|^nine|^ten", title, ignore.case=TRUE)
> d.power <- 
>   d %>%
>   select( title, claps, clap.score ) %>%
>   filter( power.list )
> ```
> 
> You could take it a step further by isolating the messy regular expression:
> 
> ```r
> regex.power <- 
>   "^[[:digit:]]|^one|^two|^three|^four|
>    ^five|^six|^seven|^eight|^nine|^ten"
> 
> power.list <- 
>   grepl( regex.power, title, ignore.case=TRUE )
> 
> d.power <- 
>   d %>%
>   select( title, claps, clap.score ) %>%
>   filter( power.list )
> ```
> 
> It makes each step easier to see, makes errors a lot easier to identify. The goal is not the least number of lines, but the most readable and easy-to-maintain code.

@lecy Got it! 

This helps a lot to see the differences, and to see different ways of writing more readable code especially as we progress through the course.

Thank you for taking the time to show more examples!

**Comment by Mazhar3204** [comment]
Hi Professor, 

Is this a way to do it?

clap.score <- log( d$clap + 1 )
How.to <- grepl( "How", d$title,ignore.case=TRUE)

d.how <- 
  d %>%
  select( title, claps) %>%
  filter( How.to)

mean(d.how$claps)

**Comment by lecy** [comment]
You can approach it in a couple of ways - keeping all of the data together in a dataframe, or isolating the one stat you need. 

## Data Frame Approach

```r
# make sure you add the score back to the df
d$clap.score <- log( d$claps + 1 )
d$power.list <- grepl( ... )

d %>%
  group_by( power.list ) %>% 
  summarize( ave=mean( clap.score )
```

## Vector Approach 

```r
# isolate two vectors needed
clap.score <- log( d$claps + 1 )
power.list <- grepl( ... )
clap.score[ power.list ] %>% mean()
```

Both are fine here - typically it will mostly depend on your work flow. Do you just need the stat (vectors are easy), or do you need to keep all of the variables together in a single data frame to use for subsequent steps (data frame approach might be better).
